File: SuperHero/app/src/main/java/pe/isil/superhero/HeroRepository.java
package pe.isil.superhero;

import android.util.Log;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.Objects;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class HeroRepository {

    private ShowHeroesInterface showHeroesInterface;

    public void setShowHeroesInterface(ShowHeroesInterface showHeroesInterface) {
        this.showHeroesInterface = showHeroesInterface;
    }

    public void searchHeroes(String heroName) {

        Retrofit retrofit = ApiClient.getRetrofit();
        HeroService heroService = retrofit.create(HeroService.class);

        Call<HeroResponse> searchHero = heroService.searchHero(heroName);

        searchHero.enqueue(new Callback<HeroResponse>() {
            @Override
            public void onResponse(@NonNull Call<HeroResponse> call, @NonNull Response<HeroResponse> response) {
                if (response.isSuccessful()){
                    HeroResponse heroResponse = response.body();
                    showHeroesInterface.showHeroes(Objects.requireNonNull(heroResponse).getHeroes());
                }
            }

            @Override
            public void onFailure(@NonNull Call<HeroResponse> call, Throwable t) {
                Log.d("HeroRepository", t.toString());
            }
        });
    }

    public interface ShowHeroesInterface {
        void showHeroes(ArrayList<Hero> heroes);
    }
}

File: SuperHeroCompose/app/src/main/java/pe/isil/superherocompose/repository/HeroRepository.kt
package pe.isil.superherocompose.repository

import android.util.Log
import pe.isil.superherocompose.data.model.Hero
import pe.isil.superherocompose.data.model.HeroResponse
import pe.isil.superherocompose.data.remote.ApiClient
import pe.isil.superherocompose.data.remote.HeroService
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class HeroRepository(

    private val heroService: HeroService = ApiClient.getHeroService()
) {
    fun searchHero(name: String, callback: (ArrayList<Hero>) -> Unit){

        val searchHero = heroService.searchHero(name)

        searchHero.enqueue(object: Callback<HeroResponse> {
            override fun onResponse(call: Call<HeroResponse>, response: Response<HeroResponse>) {
                if (response.isSuccessful) {
                    callback((response.body() as HeroResponse).heroes)
                }
            }

            override fun onFailure(call: Call<HeroResponse>, t: Throwable) {
                t.message?.let { Log.d("HeroRepository", it) }
            }
        })

    }
}